From e05fc17118b6dc20adfe0652756df023528a7fc0 Mon Sep 17 00:00:00 2001
From: Jesper Lloyd <jpl.lloyd@gmail.com>
Date: Sat, 7 Sep 2019 09:38:28 +0200
Subject: [PATCH] Alter build config to work with swig >= 4

A change in the generated module import makes builds
created with swig >= 4 fail to import mypaintlib.
This changes the installation layout to handle that.
The relevant swig change may have been introduced
in 3.0.9, but I have not verified this.

Note: In order to test-run from the source directory,
create a symlink to _mypaintlib.so in lib.
E.g (from repo root): ln -s ../_mypaintlib.so lib/_mypaintlib.so

For the above reason, lib/_mypaintlib.so is no longer removed
by scons - I think we can safely assume that anyone building
this branch will not have ancient build artifacts lying around.
---
 SConscript          | 10 +++++-----
 SConstruct          |  1 -
 gui/autorecover.py  |  2 +-
 gui/layerswindow.py |  3 +--
 4 files changed, 7 insertions(+), 9 deletions(-)

diff --git a/SConscript b/SConscript
index a5edfb0bb..85534aa4c 100644
--- a/SConscript
+++ b/SConscript
@@ -122,16 +122,16 @@ install_perms(env, '$prefix/share/applications', 'desktop/mypaint.desktop')
 install_perms(env, '$prefix/share/appdata', 'desktop/mypaint.appdata.xml')
 
 # location for achitecture-dependent modules
-install_perms(env, '$prefix/lib/mypaint', mypaintlib)
+install_perms(env, '$prefix/lib/mypaint/lib/', mypaintlib)
 
 # Program and supporting UI XML
 install_perms(env, '$prefix/bin', 'mypaint', perms=0755)
 install_perms(env, '$prefix/share/mypaint/gui', Glob('gui/*.xml'))
 install_perms(env, '$prefix/share/mypaint/gui', Glob('gui/*.glade'))
-install_perms(env, "$prefix/share/mypaint/lib",      Glob("lib/*.py"))
-install_perms(env, "$prefix/share/mypaint/lib/layer", Glob("lib/layer/*.py"))
-install_perms(env, "$prefix/share/mypaint/gui",      Glob("gui/*.py"))
-install_perms(env, "$prefix/share/mypaint/gui/colors", Glob("gui/colors/*.py"))
+install_perms(env, "$prefix/lib/mypaint/lib",      Glob("lib/*.py"))
+install_perms(env, "$prefix/lib/mypaint/lib/layer", Glob("lib/layer/*.py"))
+install_perms(env, "$prefix/lib/mypaint/gui",      Glob("gui/*.py"))
+install_perms(env, "$prefix/lib/mypaint/gui/colors", Glob("gui/colors/*.py"))
 
 
 Return('mypaintlib')
diff --git a/SConstruct b/SConstruct
index 6dc6c5041..48c022daf 100644
--- a/SConstruct
+++ b/SConstruct
@@ -141,7 +141,6 @@ env.Execute(Delete([
     'libmypaintlib.so',
     'libmypaint.a',
     'libmypaint-tests.a',
-    'lib/_mypaintlib.so',
     ]))
 
 
diff --git a/gui/autorecover.py b/gui/autorecover.py
index 615aba0ea..f700b03d6 100644
--- a/gui/autorecover.py
+++ b/gui/autorecover.py
@@ -44,7 +44,7 @@ def __init__(self, app, **kwargs):
         """Initialize for the main app and its working doc."""
         super(Presenter, self).__init__(**kwargs)
         self._app = weakref.proxy(app)
-        builder_xml = os.path.splitext(__file__)[0] + ".glade"
+        builder_xml = os.path.join(self._app.datapath, "gui/autorecover.glade")
         builder = Gtk.Builder()
         builder.set_translation_domain("mypaint")
         builder.add_from_file(builder_xml)
diff --git a/gui/layerswindow.py b/gui/layerswindow.py
index a1a52f249..0607106e9 100644
--- a/gui/layerswindow.py
+++ b/gui/layerswindow.py
@@ -123,8 +123,7 @@ def __init__(self):
         view_scroll.set_hexpand(True)
         view_scroll.set_vexpand(True)
         # Context menu
-        ui_dir = os.path.dirname(os.path.abspath(__file__))
-        ui_path = os.path.join(ui_dir, "layerswindow.xml")
+        ui_path = os.path.join(self.app.datapath, "gui/layerswindow.xml")
         self.app.ui_manager.add_ui_from_file(ui_path)
         menu = self.app.ui_manager.get_widget("/LayersWindowPopup")
         menu.set_title(_("Layer"))
